Fibaro components

pbalogh77 opened the pull request:
## Description:
Adds cover, light, sensor and switch support to the Fibaro component

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pbalogh77 committed:
Added Fibaro omcponents

Added cover, light, sensor and switch components

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def __init__(self, fibaro_device, controller):
+        """Initialize the Vera device."""
+        FibaroDevice.__init__(self, fibaro_device, controller)
```
You're inheriting from `FibaroDevice`, so you can just do:

```
super().__init__(...)
```

pbalogh77 replied:
Ok, thanks

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def set_cover_tilt_position(self, **kwargs):
+        """Move the cover to a specific position."""
+        self.action("setValue2", kwargs.get(ATTR_TILT_POSITION))
```
Don't use a string literal here – put a constant in its place.

pbalogh77 replied:
I agree partially as this action is invoked from multiple locations. However, I think I will rather add it as a function, to have it simpler and easier to understand.

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def open_cover(self, **kwargs):
+        """Open the cover."""
+        self.action("open")
```
See above.

pbalogh77 replied:
With all due respect, I don't fully agree with you here. Having a command string which is used from a single place in all of home assistant to be moved from that single place, placed in a file that does not use them and import it. It not only makes the source bigger, but also reduces clarity, at least in my opinion.
It's used from the very file which is responsible for this specific functionality, which seems like the ideal place for it.

pbalogh77 replied:
Thanks :-)

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def close_cover(self, **kwargs):
+        """Close the cover."""
+        self.action("close")
```
See above.

pbalogh77 replied:
With all due respect, I don't fully agree with you here. Having a command string which is used from a single place in all of home assistant to be moved from that single place, placed in a file that does not use them and import it. It not only makes the source bigger, but also reduces clarity, at least in my opinion.
It's used from the very file which is responsible for this specific functionality, which seems like the ideal place for it.

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def open_cover_tilt(self, **kwargs):
+        """Open the cover tilt."""
+        self.action("setValue2", 100)
```
See above.

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def close_cover_tilt(self, **kwargs):
+        """Close the cover."""
+        self.action("setValue2", 0)
```
See above.

bachya commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+
+    def stop_cover(self, **kwargs):
+        """Stop the cover."""
+        self.action("stop")
```
See above.

pbalogh77 replied:
With all due respect, I don't fully agree with you here. Having a command string which is used from a single place in all of home assistant to be moved from that single place, placed in a file that does not use them and import it. It not only makes the source bigger, but also reduces clarity, at least in my opinion.
It's used from the very file which is responsible for this specific functionality, which seems like the ideal place for it.

bachya commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,184 @@
+                # so instead, we switch it off
+                elif target_brightness < 4:
+                    self._brightness = 0
+                    self.action("turnOff")
```
See above.

bachya commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,184 @@
+                               int(rgb[2] * self._brightness / 255.0 + 0.5),
+                               int(self._white * self._brightness / 255.0 +
+                                   0.5))
+                if self.state == 'off':
```
See above.

bachya commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,184 @@
+                    self._brightness and self._brightness >= 4:
+                self._last_brightness = self._brightness
+            self._brightness = 0
+            self.action("turnOff")
```
See above.

bachya commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+        self.last_changed_time = None
+        FibaroDevice.__init__(self, fibaro_device, controller)
+        self.entity_id = ENTITY_ID_FORMAT.format(self.ha_id)
+        if fibaro_device.type in SENSOR_TYPES.keys():
```
```
if fibaro_device.type in SENSOR_TYPES:
```

pbalogh77 replied:
Ok, thanks, will fix

bachya commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+            if self.fibaro_device.properties.unit == 'F':
+                return TEMP_FAHRENHEIT
+            return self.fibaro_device.properties.unit
+        except (KeyError, ValueError):
```
When would these errors actually occur?

bachya commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+            return self.fibaro_device.properties.unit
+        except (KeyError, ValueError):
+            pass
+        return ''
```
```
return None
```

pbalogh77 replied:
Thanks, done

bachya commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+        """Update the state."""
+        try:
+            self.current_value = float(self.fibaro_device.properties.value)
+        except (KeyError, ValueError):
```
When would these errors actually occur?

bachya commented on homeassistant/components/switch/fibaro.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, fibaro_device, controller):
+        """Initialize the Fibaro device."""
+        self._state = False
+        FibaroDevice.__init__(self, fibaro_device, controller)
```
You're inheriting from `FibaroDevice`, so you can just do:

```
super().__init__(...)
```

bachya commented on homeassistant/components/switch/fibaro.py:
```diff
@@ -0,0 +1,68 @@
+
+    def turn_on(self, **kwargs):
+        """Turn device on."""
+        self.action("turnOn")
```
See above.

bachya commented on homeassistant/components/switch/fibaro.py:
```diff
@@ -0,0 +1,68 @@
+
+    def turn_off(self, **kwargs):
+        """Turn device off."""
+        self.action("turnOff")
```
See above.

pbalogh77 committed:
Improvements based on code review

Improvements based on code review

amelchio commented:
Just a note, it is probably better to create individual PRs for each of the platforms. That way smaller parts can get approved and merged faster without a single person having to read, understand and approve all of it.

That said, I think you should keep this PR now as it is not huge and review has already started.

pbalogh77 commented:
Ooops, clicked the wrong button when commenting :-( and I accidentally closed it...

amelchio commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,185 @@
+        with self._update_lock:
+            if self._supported_flags & SUPPORT_BRIGHTNESS:
+                target_brightness = None
+                if ATTR_BRIGHTNESS_PCT in kwargs:
```
This never happens, the light component does the percent -> byte conversion for you. So only `ATTR_BRIGHTNESS` is passed to your light platform.

amelchio commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,185 @@
+                        else:
+                            self._brightness = 255
+                # We're supposed to turn it on to a very very low level,
+                # so instead, we switch it off
```
How do we set it very very low then?

I suggest you remove all the magic constants throughout and just set it to the requested value.

amelchio replied:
Those are UI concerns. The light platform should just do as it's told :)

(We actually changed this in the UI recently, home-assistant/home-assistant-polymer#1715, because it is frustrating to be left in the dark when trying to set a low brightness)

pbalogh77 replied:
I see and this is a very valid point from a system perspective. Should the remembering of previous brightness level also be done by HA? Currently I implemented it, as Fibaro does it from its UI but not from its API.

amelchio commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,185 @@
+                # so instead, we switch it off
+                elif target_brightness < 4:
+                    self._brightness = 0
+                    self.call_turn_off()
```
Do not turn off in `turn_on()` ;-)

amelchio commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+        else:
+            self._icon = None
+            self._device_class = None
+            self._unit_of_measurement = None
```
Please order the same way with if/else, and preferably in the same order as in `SENSOR_TYPES`. Makes it easier to read :)

pbalogh77 replied:
Done. Thanks!

amelchio commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+
+    @property
+    def state(self):
+        """Return the name of the sensor."""
```
Typo in comment, name -> state

pbalogh77 replied:
Done, thanks

amelchio commented on homeassistant/components/sensor/fibaro.py:
```diff
@@ -0,0 +1,99 @@
+        try:
+            if self._unit_of_measurement:
+                return self._unit_of_measurement
+            if self.fibaro_device.properties.unit == 'lux':
```
Is this information available in `__init__()` so we can find it just once?

pbalogh77 replied:
Good idea, I think it'll work and it's much nicer. Thanks

amelchio reviewed (changes requested):
Looks mostly good.

The thresholds in the light platform seem strange and complicate things. Is there some hardware reason that these are needed?

pbalogh77 commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+    @property
+    def current_cover_position(self):
+        """Return current position of cover. 0 is closed, 100 is open."""
+        return self.bound(self.get_level())
```
It doesn't do IO, just reads a variable. I will turn the get_level of the FibaroDevice into a property as well, so it's more trivial what happens here.

amelchio replied:
IMO it's fine as is.

pbalogh77 replied:
Did the rework nevertheless :-)

pbalogh77 commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+    @property
+    def is_closed(self):
+        """Return if the cover is closed."""
+        if self.current_cover_position is not None:
```
Ok, thanks!

andre-richter commented:
Hi @pbalogh77, 

I am using quite a lot of the Fibaro roller shutters and switches with HA, but I am still quite uneducated about HA source code.
Could you explain to me what your PR will change with regards to the status quo? I think I see blind tilt support. Also refactoring to some new HA code base?

Thanks a lot for your efforts!

Andre

pbalogh77 commented on homeassistant/components/cover/fibaro.py:
```diff
@@ -0,0 +1,91 @@
+import logging
+
+from homeassistant.components.cover import CoverDevice, ENTITY_ID_FORMAT, \
+    ATTR_POSITION, ATTR_TILT_POSITION
```
Thanks, done.

pbalogh77 commented:
> Hi @pbalogh77,
> 
> I am using quite a lot of the Fibaro roller shutters and switches with HA, but I am still quite uneducated about HA source code.
> Could you explain to me what your PR will change with regards to the status quo? I think I see blind tilt support. Also refactoring to some new HA code base?
> 
> Thanks a lot for your efforts!
> 
> Andre

Hi Andre,

It adds support for Fibaro Home Center hubs, so instead of connecting the Z-wave devices to your HA with a Z-wave stick, you can now use the Fibaro hub and integrate that. The reason for this is that while HA is great, Fibaro HC2 is rock solid and has some features specific to it. I also believe that HA is best at integrating hubs, e.g. not taking over the responsibility of everything, just combining functionality on a higher level. Also, this way I can have the fibaro app as a fall-back if HA stops or crashes to still get into my home :-) In this way, I think HA can be used as a alternative UI for Fibaro hub users too.

Cheers,
Peter

pbalogh77 committed:
Fixes based on code review

Fixes based on code review

andre-richter commented:
> > Hi @pbalogh77,
> > I am using quite a lot of the Fibaro roller shutters and switches with HA, but I am still quite uneducated about HA source code.
> > Could you explain to me what your PR will change with regards to the status quo? I think I see blind tilt support. Also refactoring to some new HA code base?
> > Thanks a lot for your efforts!
> > Andre
> 
> Hi Andre,
> 
> It adds support for Fibaro Home Center hubs, so instead of connecting the Z-wave devices to your HA with a Z-wave stick, you can now use the Fibaro hub and integrate that. The reason for this is that while HA is great, Fibaro HC2 is rock solid and has some features specific to it. I also believe that HA is best at integrating hubs, e.g. not taking over the responsibility of everything, just combining functionality on a higher level. Also, this way I can have the fibaro app as a fall-back if HA stops or crashes to still get into my home :-) In this way, I think HA can be used as a alternative UI for Fibaro hub users too.
> 
> Cheers,
> Peter

Ah, understand! I am using HA without the Fibaro home center so far, and it works quite well. I will keep this in mind though.

Thanks!

pbalogh77 committed:
Changes to light behavior based on code review

Changes to light behavior based on code review

pbalogh77 committed:
Internal changes

Changed how brightness is represented internally.
It should have no impact on functionality.

amelchio commented:
Let's merge this (when the build passes) so the new Fibaro component is complete when it gets released. @pbalogh77 has promised in chat to add documentation soon.

MartinHjelmare commented on homeassistant/components/light/fibaro.py:
```diff
@@ -0,0 +1,165 @@
+
+    def turn_on(self, **kwargs):
+        """Turn the light on."""
+        with self._update_lock:
```
This could block threads in the thread pool. That's not allowed. A non blocking acquire could work. What are we trying to protect?

pbalogh77 replied:
The issue I'm trying to solve is that while we're still executing the turn_on command, another thread is receiving notifications from the fibaro hub and processing them, e.g. there can be calls to the update function. Which can mess up the data a bit and cause some havoc. This is especially troubling in the case of a turned off colored light, which has to issue two consecutive rest api calls to the fibaro hub, both of them resulting one or multiple status updates from the fibaro hub. Since neither function has infinite loops and only of them has actual IO, I don't see how it could lock up a thread permanently, but I'm very open to other ideas of solving this race condition.

amelchio replied:
The race is real, I had the same issue with LIFX. I am not sure how to best solve it for a threaded platform. As is, I guess it could get ugly if the hub gets unavailable.

balloob replied:
So this should become `async_turn_on` and Fibaro needs to maintain an async lock. We can't solve this on the Home Assistant level as we don't know exactly how the resources need to be treated that need locking (is it per device, per hub etc).

amelchio commented:
I’ll do a judgement call and merge this so it makes the beta along with the remaining Fibaro news.

I am confident that @pbalogh77 is willing to address the concern that Martin raised in a separate PR, in time for the final release.

balloob commented:
Good call 👍

balloob commented:
I am stoked. I think that Fibaro has done a great job building a stable platform, with support for their hub, we will be accessible to a new group of people.

MartinHjelmare commented:
If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion.

Thanks!
